audit/generic_auditor.py

```diff
@@ -0,0 +1,652 @@
+
+        # Filter out any participants with no samples and log any found
+        for participant in participant_data:
+            if not participant.get('samples'):
+                logging.info(
+                    f'{self.dataset} :: Filtering participant {participant.get("id")} ({participant.get("externalId")}) it has no samples.'
+                )
+        participant_data = [
+            participant
+            for participant in participant_data
+            if participant.get('samples')
+        ]
+
+        return participant_data
+
+    @staticmethod
+    def map_participants_to_samples(
+        participants: list[dict],
+    ) -> dict[ParticipantExternalId, list[SampleId]]:
+        """
+        Returns the {external_participant_id : sample_internal_id} mapping for a Metamist dataset
+        - Also reports if any participants have more than one sample
+        """
+
+        # Create mapping of participant external ID to sample ID
+        participant_eid_sample_id_map = {
+            participant.get('externalId'): [
+                sample.get('id') for sample in participant.get('samples')
+            ]
+            for participant in participants
+        }
+
+        # Create mapping of participant external ID to internal ID
+        participant_eid_to_iid = {
+            participant.get('externalId'): participant.get('id')
+            for participant in participants
+        }
+
+        # Report if any participants have more than one sample ID
+        for participant_eid, sample_ids in participant_eid_sample_id_map.items():
+            if len(sample_ids) > 1:
+                logging.info(
+                    f'Participant {participant_eid_to_iid[participant_eid]} (external ID: {participant_eid})'
+                    + f' associated with more than one sample id: {sample_ids}'
+                )
+
+        return participant_eid_sample_id_map
+
+    @staticmethod
+    def map_internal_to_external_sample_ids(
+        participants: list[dict],
+    ) -> dict[SampleId, SampleExternalId]:
+        """
+        Returns the {internal sample ID : external sample ID} mapping for all participants in a Metamist dataset
+        Also removes any samples that are part of the exclusions list
+        """
+        # Create a list of dictionaries, each mapping a sample ID to its external ID
+        samples_all = [
+            {
+                sample.get('id'): sample.get('externalId')
+                for sample in participant['samples']
+            }
+            for participant in participants
+        ]
+
+        # Squish them into a single dictionary not a list of dictionaries
+        sample_internal_external_id_map = {
+            sample_id: sample_external_id
+            for sample_map in samples_all
+            for sample_id, sample_external_id in sample_map.items()
+            if sample_id not in EXCLUDED_SAMPLES
+        }
+
+        return sample_internal_external_id_map
+
+    def get_sequence_mapping(
```

Comment:
there are a lot of `continue`s in this block; are there known metamist results we expect to trigger all these failure conditions, or is this just super cautious?

Reply:
If you want to soft-succeed (not sure that's a phrase) this is fine, if you want to hard fail when you encounter wacky data it might be worth appending all failure types to a list, then at the end of the method you can decide whether to return the data (no failures) or raise an exception (and print the failures)

Reply:
Yeah this function borrows a lot from the `find_sequence_files_to_delete` script [here](https://github.com/populationgenomics/sample-metadata/blob/72b5ea46bf73c952119bfb63990c9f55b0f7f228/scripts/find_sequence_files_to_delete.py#L69). I believe there are some older sequence entries with odd data formatting, its overly cautious now but can probably be toned back in future as new sequence/assay entries follow a standardised format